x-pack/plugins/apm/server/lib/traces/get_top_traces.js

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import moment from 'moment';
+import {
+  TRANSACTION_NAME,
+  TRANSACTION_ID,
+  TRANSACTION_DURATION
+} from '../../../common/constants';
+import { get, sortBy } from 'lodash';
+
+export async function getTopTraces({ setup }) {
+  const { start, end, esFilterQuery, client, config } = setup;
+
+  const duration = moment.duration(end - start);
+  const minutes = duration.asMinutes();
+
+  const params = {
+    index: config.get('apm_oss.transactionIndices'),
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          must: {
+            exists: {
+              field: 'trace.id' // todo: move to constant? or TS?
+            }
+          },
+          must_not: {
+            exists: {
+              field: 'parent.id' // todo: move to constant? or TS?
+            }
+          },
+          filter: [
+            {
+              range: {
+                '@timestamp': {
+                  gte: start,
+                  lte: end,
+                  format: 'epoch_millis'
+                }
+              }
+            }
+          ]
+        }
+      },
+      aggs: {
+        transactions: {
+          terms: {
+            field: `${TRANSACTION_NAME}.keyword`,
+            order: { avg: 'desc' },
+            size: 100
+          },
+          aggs: {
+            sample: {
+              top_hits: {
+                _source: [TRANSACTION_ID, 'context.service.name'],
+                size: 1,
+                sort: [{ '@timestamp': { order: 'desc' } }]
+              }
+            },
+            avg: {
+              avg: { field: TRANSACTION_DURATION }
+            },
+            p95: {
+              percentiles: {
+                field: TRANSACTION_DURATION,
+                percents: [95]
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  if (esFilterQuery) {
+    params.body.query.bool.filter.push(esFilterQuery);
+  }
+
+  const resp = await client('search', params);
+  const buckets = get(resp, 'aggregations.transactions.buckets', []);
+  const results = buckets.map(bucket => {
+    const averageResponseTime = bucket.avg.value;
+    const tracesPerMinute = bucket.doc_count / minutes;
+    const impact = Math.round(averageResponseTime * tracesPerMinute);
+    const sample = get(bucket, 'sample.hits.hits[0]._source');
+    return {
+      name: bucket.key,
+      serviceName: get(sample, 'context.service.name', 'n/a'),
+      id: get(sample, TRANSACTION_ID),
+      p95: bucket.p95.values['95.0'],
+      averageResponseTime,
+      tracesPerMinute,
+      impact
+    };
+  });
+
+  // Sort results by impact - needs to be desc, hence the reverse()
+  return sortBy(results, o => o.impact).reverse();
```

Comment:
We do actually have _.orderBy that supports sort order:

https://github.com/elastic/kibana/blob/4126d51044d8e85f80548e21e1615d20ded715d8/x-pack/package.json#L148

Reply:
Ah yeah this was just all moved over from the get_top_transactions file

https://github.com/elastic/kibana/blob/2c4e263982bd0517a1de4eda970898f76c9e5aed/x-pack/plugins/apm/server/lib/transactions/get_top_transactions.js#L100

I was considering making "get top traces" (maybe renamed "get trace list") to just use get transactions under the hood, with maybe an option to adjust the query slightly, since so much of those files are now duplicate. Thoughts?

Reply:
Ah, I love pointing out my own mistakes in other people's PRs :D